### PR TITLE
common: Improving message sent to user when getting signals

### PR DESCRIPTION
**Pull request by ErwanAliasr1**
As per https://tracker.ceph.com/issues/23320, if we get an abort() or pthread_kill() we could end up in a situation were the info reported to the user is very misleading since we try to get the program name while pid = 0.
In this particular case, we spent time trying to understand the actual value of the structure and why it was reported as 0/Unknown.

The fact this is because the kernel is the emitter in such case. So this PR is about making the sender being explicted when its kernel and offer a way to specify a proper message regarding si_code value, unless print the structure for debugging purpose.

### CODE REVIEW

**Comment by ErwanAliasr1** [comment]
Rebased.

File: src/global/signal_handler.cc [hunk 1]
```diff
@@ -313,13 +313,33 @@ struct SignalHandler : public Thread {
...
-		   << " from " << " PID: " << siginfo->si_pid
-		   << " task name: " << task_name
-		   << " UID: " << siginfo->si_uid
-		   << dendl;
-	      handlers[signum]->handler(signum);
+              ostringstream message;
+              message << "received  signal: " << sig_str(signum);
+              switch (siginfo->si_code) {
+                case SI_USER:
+                  message << " from " << get_name_by_pid(siginfo->si_pid);
```
Comment (tchaikov) [inline src/global/signal_handler.cc]: i'd suggest update `get_name_by_pid()`, and return "kernel" if `pid` is 0. please note, Ceph is portable and runs on  FreeBSD also.

File: src/global/signal_handler.cc [hunk 2]
```diff
@@ -313,13 +313,33 @@ struct SignalHandler : public Thread {
...
+                  // If PID is undefined, it doesn't have a meaning to be displayed
+                  if (siginfo->si_pid) {
+                    message << " (PID: " << siginfo->si_pid << ")";
+                  } else {
+                    message << " ( Could be generated by pthread_kill(), raise(), abort(), alarm() )";
```
Comment (tchaikov) [inline src/global/signal_handler.cc]: it does not help from user's perspective. probably we can just print out the PID even it's zero.

File: src/global/signal_handler.cc [hunk 3]
```diff
@@ -313,13 +313,33 @@ struct SignalHandler : public Thread {
...
+                  }
+                  message << " UID: " << siginfo->si_uid;
+                  break;
+                default:
+                  /* As we have an unhandled signal, let's report the structure to help debugging */
```
Comment (tchaikov) [inline src/global/signal_handler.cc]: i don't think the signal is unhandled.

File: src/global/signal_handler.cc [hunk 4]
```diff
@@ -313,13 +313,33 @@ struct SignalHandler : public Thread {
...
+                  message << ", si_code : " << siginfo->si_code;
+                  message << ", si_value (int): " << siginfo->si_value.sival_int;
+                  message << ", si_value (ptr): " << siginfo->si_value.sival_ptr;
+                  message << ", si_errno: " << siginfo->si_errno;
+                  message << ", si_pid : " << siginfo->si_pid;
+                  message << ", si_uid : " << siginfo->si_uid;
+                  message << ", si_addr" << siginfo->si_addr;
+                  message << ", si_status" << siginfo->si_status;
```
Comment (tchaikov) [inline src/global/signal_handler.cc]: i'd suggest not print these variables unless they will be helpful for postmortem debugging.

File: src/global/signal_handler.cc [hunk 5]
```diff
@@ -313,13 +313,33 @@ struct SignalHandler : public Thread {
...
 	    if (r == 1) {
 	      siginfo_t * siginfo = &handlers[signum]->info_t;
-	      string task_name = get_name_by_pid(siginfo->si_pid);
-	      derr << "received  signal: " << sig_str(signum)
-		   << " from " << " PID: " << siginfo->si_pid
-		   << " task name: " << task_name
-		   << " UID: " << siginfo->si_uid
-		   << dendl;
-	      handlers[signum]->handler(signum);
+              ostringstream message;
```
[the remark this answers is not in the document]
Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: That was my first idea but figured that derr expects a dendl for each line making this single line output generated in multi line of code impossible. Am I wrong ?

File: src/global/signal_handler.cc (hunk 2 above)
Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: I put that as it took me some time to figure that this signal could be generated by the code itself and no only by a ctrl+c or explicit kill.

File: src/global/signal_handler.cc (hunk 3 above)
Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: Right. Updated this way.

File: src/global/signal_handler.cc (hunk 4 above)
Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: That's exactly the purpose of that part. When I had to debug a issue related to this code, I had to print all that structure as fields depends on the signaled as per : https://www.mkssoftware.com/docs/man5/siginfo_t.5.asp#Signal_Codes

File: src/global/signal_handler.cc (hunk 1 above)
Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: That's exactly the aim of the first patch of this PR.
I updated it to return "Kernel" instead of "Linux Kernel".

**Comment by ErwanAliasr1** [comment]
@tchaikov Is it better this way ?

File: src/global/signal_handler.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/global/signal_handler.cc]: no, you are right. i just thought it's overkill to print them in a single line. and if you insist, you could use `*_dout` .

File: src/global/signal_handler.cc (hunk 4 above)
Comment (tchaikov) [reply src/global/signal_handler.cc]: @ErwanAliasr1 i don't really understand. if your intention is to debugging an issue related to this code, i guess you've finished debugging, no?

File: src/global/signal_handler.cc (hunk 2 above)
Comment (tchaikov) [reply src/global/signal_handler.cc]: well, how do we expect user to interpret this ?

> ( Could be generated by pthread_kill(), raise(), abort(), alarm() )

Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: It's not really for users but much more for advanced users or support.
I was in a support case where this code got triggered and we though for a while that the signal was sent by an external process as the error message was saying so.
It took sometime before considering the signal was not sent by an external tool but self-triggered by ceph itself.
If I had that message saying the signal could be triggered by ceph itself then I could have being focused on the real cause and not tracking my system & other processes.
That's why adding this information could save time for the next people facing this situation.

File: src/global/signal_handler.cc (hunk 4 above)
Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: Looks like I'm not clear enough :)
During my debugging sessions, If I had that information being printed, I would have been in a better position to understand what type of signal we got & its meaning.
So my debbugging sessions was to print that information to sort out the case we were in, meaning, checkouting the tag, adding this patch, rebuilding & trick the versionning to let this binary being accepted by the other libs.
That's a lot of energy for very few information.
This code isn't supposed to be printed unless we are in a weird case. If so, printing will save a lot of time for the support/developper.

File: src/global/signal_handler.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (ErwanAliasr1) [reply src/global/signal_handler.cc]: but isn't dout to be on stdout while I need stderr ?

**Comment by ErwanAliasr1** [comment]
Rebased against master

**Comment by ErwanAliasr1** [comment]
Oh sorry @smithfarm @tchaikov ... I missed that one.

Fixed & rebased.

**Comment by smithfarm** [comment]
@ErwanAliasr1 thanks!

**Comment by ErwanAliasr1** [comment]
Rebased against master.
@tchaikov is it ok for you then ?

**Comment by ErwanAliasr1** [comment]
Rebased on master to stay up to date

**Comment by ErwanAliasr1** [comment]
Rebased against master.

**Comment by ErwanAliasr1** [comment]
Can we make this PR being merged ?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-04-23_15:07:57-rados-wip-sage3-testing-2018-04-23-0831-distro-basic-smithi/

**Pull request merged**